### PR TITLE
util: add type checking to CONTAINER_OF

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Add a SAME_TYPE checking macro and use it to add type validation to CONTAINER_OF. This is inspired by the Linux container_of implementation.

The check is inhibited when using C++ as __builtin_types_compatible_p is not available there.

---

Linux version: https://elixir.bootlin.com/linux/latest/source/include/linux/container_of.h#L18

Deps:
- https://github.com/zephyrproject-rtos/zephyr/pull/61936
- https://github.com/zephyrproject-rtos/zephyr/pull/62023
- https://github.com/zephyrproject-rtos/zephyr/pull/62024
- https://github.com/zephyrproject-rtos/zephyr/pull/62018
- https://github.com/zephyrproject-rtos/zephyr/pull/62019
- https://github.com/zephyrproject-rtos/zephyr/pull/62020
- https://github.com/zephyrproject-rtos/zephyr/pull/62021
- https://github.com/zephyrproject-rtos/zephyr/pull/62022
- https://github.com/thesofproject/sof/pull/8112
  - https://github.com/zephyrproject-rtos/zephyr/pull/61991
  - https://github.com/zephyrproject-rtos/zephyr/pull/62262
- https://github.com/zephyrproject-rtos/zephyr/pull/62246

Push test run: https://github.com/zephyrproject-rtos/zephyr-testing/tree/vfabio-testing-branch

^^ there's some unrelated breakage in this one, but no type checking errors.

There may be more issues in the weekly CI run, but this is as much as I could fix before that.

### CODE REVIEW

**Review by andyross — APPROVED** [review]
Nice.  I'm actually surprised this isn't finding more problems.

**Comment by fabiobaltieri** [comment]
> Nice. I'm actually surprised this isn't finding more problems.

Yeah it wasn't too bad, sure there will be more in the weekly run.

File: include/zephyr/sys/util.h [hunk 1]
```diff
@@ -201,6 +202,27 @@ extern "C" {
...
+ * @param a the first entity to be compared
+ * @param a the second entity to be compared
+ * @return 1 if the two elements are compatible, 0 if they are not
+ */
+#define SAME_TYPE(a, b) __builtin_types_compatible_p(__typeof__(a), __typeof__(b))
+
+/**
+ * @brief Validate CONTAINER_OF parameters, only applies to C mode.
+ */
+#ifndef __cplusplus
```
Comment (keith-zephyr) [inline include/zephyr/sys/util.h:217]: Should this also check `HAS_BUILTIN(__builtin_types_compatible_p)`.  Do you know if __builtin_types_compatible_p is supported by all toolchains.  I know GCC and clang definitely has support.

**Review by keith-zephyr — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

File: include/zephyr/sys/util.h [hunk 2]
```diff
@@ -15,6 +15,7 @@
 #define ZEPHYR_INCLUDE_SYS_UTIL_H_
 
 #include <zephyr/sys/util_macro.h>
+#include <zephyr/toolchain/common.h>
```
Comment (marc-hb) [inline include/zephyr/sys/util.h:18]: This addition just revealed a funny ZRESTRICT conflict between `include/zephyr/toolchain/common.h` and `include/zephyr/toolchain/gcc.h` when compiling C++

- #27245

As a datapoint/evidence, replacing this line with `#include <zephyr/toolchain/gcc.h>` avoids the warning.

https://github.com/thesofproject/sof/actions/runs/6127039819/job/16632163797?pr=8175


```
/zep_workspace/sof/src/audio/module_adapter/iadk/module_initial_settings_concrete.cpp
In file included from /zep_workspace/zephyr/include/zephyr/toolchain.h:50,
                 from /zep_workspace/zephyr/include/zephyr/sys/time_units.h:10,
                 from /zep_workspace/zephyr/include/zephyr/sys/util.h:640,
                 from /zep_workspace/sof/src/include/sof/audio/module_adapter/iadk/adsp_stddef.h:15,
                 from /zep_workspace/sof/src/audio/module_adapter/iadk/module_initial_settings_concrete.cpp:7:
/zep_workspace/zephyr/include/zephyr/toolchain/gcc.h:87: error: "ZRESTRICT" redefined [-Werror]
   87 | #define ZRESTRICT __restrict
      | 
In file included from /zep_workspace/zephyr/include/zephyr/sys/util.h:18:
/zep_workspace/zephyr/include/zephyr/toolchain/common.h:33: note: this is the location of the previous definition
   33 | #define ZRESTRICT
      | 
cc1plus: all warnings being treated as errors
```

**Comment by marc-hb** [comment]
> Nice. I'm actually surprised this isn't finding more problems.

You asked, I delivered :-)

**Review by marc-hb — COMMENTED** [review]
Not sure why this wasn't posted the first time

**Comment by marc-hb** [comment]
As ZRESTRICT does not seem to be related to CONTAINER_OF, I filed new bug
- #62464

**Pull request merged**